### PR TITLE
[time-defined cache] [follow-up]: Use a regular Mutex over RWMutex to lock the entire Get() function

### PR DESCRIPTION
**Pull request by iholder101**
### What this PR does
A quick follow-up to https://github.com/kubevirt/kubevirt/pull/11634.
Specifically, addresses @vasiliy-ul feedback in [this comment](https://github.com/kubevirt/kubevirt/pull/11634#discussion_r1555314830).

Before this PR:
The following can occur before this commit:
* thread 1 calls Get(), sees value is not set yet.
* thread 2 calls Get(), sees value is not set yet.
* thread 1 captures the write lock
* thread 2 waits on the write lock
* thread 1 calculates value and sets it, releases lock
* thread 2 catches lock, calculates value and sets it

After this PR:
Get() is being locked in its entirety and give up on using RWMutex. Not only that the behavior would be more correct, but it would also be more performant as it avoids redandunt re-calculations.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] ~Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required~
- [X] PR: The PR description is expressive enough and will help future contributors
- [X] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [X] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] ~Upgrade: Impact of this change on upgrade flows was considered and addressed if required~
- [X] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] ~Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.~
- [ ] ~Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered~

### Release note

```release-note
NONE
```

### CODE REVIEW

**Comment by iholder101** [comment]
/cc @vasiliy-ul @fossedihelm 
FYI @vladikr

File: tools/cache/time-defined-cache_test.go [hunk 1]
```diff
@@ -116,4 +117,65 @@ var _ = Describe("time defined cache", func() {
 		}).WithPolling(250 * time.Millisecond).WithTimeout(1 * time.Second)
 	})
 
+	It("should not allow two threads to get value in parallel", func() {
```
Comment (vasiliy-ul) [inline tools/cache/time-defined-cache_test.go:120]: I have some doubts about this test case. What is its purpose? Is it to test how exclusive locks work? Since it is a core golang functionality, I am not sure that we need to cover it. My main concern here is that the test is a bit hard to follow, and it seems to rely on timeouts. Are we sure it does not introduce flakiness under load?

Thoughts?

**Review by vasiliy-ul — COMMENTED** [review]
Looks good! Thanks for taking care of the issue. Only one small concern about the test (inline).

Feel free to ignore the nits if you disagree.

File: tools/cache/time-defined-cache_test.go (hunk 1 above)
Comment (iholder101) [reply tools/cache/time-defined-cache_test.go:120]: > Is it to test how exclusive locks work?

No :)
It basically tests that if multiple threads call the `Get()` method together, one thread would re-calculate and all other threads would wait and finally get the same cached value. It's valuable to guarantee this behavior. I think it's especially important when we're dealing with a cache that involves locking, as it's very easy to change implementation in a way that this behavior no longer holds.

> My main concern here is that the test is a bit hard to follow

Yeah, I agree. When go routines are involved the code is always more complex =/
However if you can think of a way to simplify it I'll be happy to hear it.

> it seems to rely on timeouts. Are we sure it does not introduce flakiness under load?

Realistically, go routines calling a function is an almost immediate operation. I don't think there's any reason for it to be delayed. I've given a 3 second timeout to be on the safe side. I don't think it's practical for it to take any longer.

WDYT?

File: tools/cache/time-defined-cache.go [hunk 2]
```diff
@@ -23,15 +23,16 @@ import (
...
+
+	k6tpointer "kubevirt.io/kubevirt/pkg/pointer"
 )
 
 type TimeDefinedCache[T any] struct {
 	minRefreshDuration time.Duration
-	lastRefresh        time.Time
-	savedValueSet      bool
+	lastRefresh        *time.Time
 	savedValue         T
```
[the remark this answers is not in the document]
Comment (iholder101) [reply tools/cache/time-defined-cache.go]: Done!

File: tools/cache/time-defined-cache.go [hunk 3]
```diff
@@ -92,6 +89,8 @@ func (t *TimeDefinedCache[T]) Set(value T) {
 
 func (t *TimeDefinedCache[T]) setWithoutLock(value T) {
 	t.savedValue = value
-	t.savedValueSet = true
-	t.lastRefresh = time.Now()
+
+	if t.lastRefresh == nil || t.minRefreshDuration.Nanoseconds() != 0 {
+		t.lastRefresh = k6tpointer.P(time.Now())
+	}
```
[the remark this answers is not in the document]
Comment (iholder101) [reply tools/cache/time-defined-cache.go:95]: I have to disagree on this one :)
That's exactly why the `pointer` package was introduced, to avoid having to introduce a variable just for the sake of having a pointer.

File: tools/cache/time-defined-cache_test.go (hunk 1 above)
Comment (vasiliy-ul) [reply tools/cache/time-defined-cache_test.go:120]: Okay, maybe let's see if we can simplify the test a bit. Added a few more comments. Do they make sense?

File: tools/cache/time-defined-cache_test.go [hunk 4]
```diff
@@ -116,4 +117,65 @@ var _ = Describe("time defined cache", func() {
...
+			getReturnValues <- ret
+		}
+
+		const goroutineCount = 5
+		for i := 0; i < goroutineCount; i++ {
+			go getValueFromCache()
+		}
+
+		// To ensure the first call is already made
+		<-firstCallMadeChannel
```
[the remark this answers is not in the document]
Comment (iholder101) [reply tools/cache/time-defined-cache_test.go]: cool!

**Comment by iholder101** [comment]
Hey @vasiliy-ul!
Thanks a lot for your valuable feedback!

I've worked on simplifying the test. I believe it's significantly simpler now. I even was able to delete a test since I figured it's redundant and the new test is good enough. PTAL.

**Comment by iholder101** [comment]
Unrelated flakieness.
/test pull-kubevirt-unit-test-arm64

**Review by vasiliy-ul — COMMENTED** [review]
:+1: A simpler test without sleeps and timeouts is much better. Looks good from my perspective. Thanks!

/approve

**Comment by iholder101** [comment]
@fossedihelm can you PTAL?

File: tools/cache/time-defined-cache_test.go [hunk 5]
```diff
@@ -71,49 +72,41 @@ var _ = Describe("time defined cache", func() {
...
+			firstCallBarrier <- struct{}{}
+			return int(recalcFunctionCallsCount), nil
 		}
 
-		cache, err := virtcache.NewTimeDefinedCache(0, true, recalcFunc)
+		cache, err := virtcache.NewTimeDefinedCache(time.Hour, true, recalcFunc)
 		Expect(err).ToNot(HaveOccurred())
 
+		const goroutineCount = 20
+		getReturnValues := make(chan int, goroutineCount)
```
Comment (fossedihelm) [inline tools/cache/time-defined-cache_test.go:89]: Even if it is closed later in the test, an Expectation could fail resulting in a non-closed channel. 
Please add a defer close (suggestion):
```
defer func() {
	_, ok := <-getReturnValues
	if ok {
		close(getReturnValues)
	}
}()
```

**Review by fossedihelm — CHANGES_REQUESTED** [review]
Looks great overall.
Some dangling open-channel

File: tools/cache/time-defined-cache_test.go [hunk 6]
```diff
@@ -71,49 +72,41 @@ var _ = Describe("time defined cache", func() {
...
 	})
 
-	It("should not allow two threads to set value in parallel", func() {
-		stopChannel := make(chan struct{})
-		defer close(stopChannel)
-		firstCallMadeChannel := make(chan struct{})
-
-		recalcFunctionCalls := int64(0)
+	It("when multiple go routines get a value only one is recalculating and others get the same cached value", func() {
+		firstCallBarrier := make(chan struct{})
```
[the remark this answers is not in the document]
Comment (iholder101) [reply tools/cache/time-defined-cache_test.go:76]: I was thinking it's necessary too, and had these in previous versions of this PR, but then I saw this from the official golang tour [1]:
> Another note: Channels aren't like files; you don't usually need to close them. Closing is only necessary when the receiver must be told there are no more values coming, such as to terminate a range loop.

If it is not necessary, it bring more damage than good because we both make and code less readable and educate to do something that's not needed.

[1] https://go.dev/tour/concurrency/4

File: tools/cache/time-defined-cache_test.go (hunk 5 above)
Comment (iholder101) [reply tools/cache/time-defined-cache_test.go:89]: see [comment above](https://github.com/kubevirt/kubevirt/pull/11663#discussion_r1557479758).

File: tools/cache/time-defined-cache_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (fossedihelm) [reply tools/cache/time-defined-cache_test.go:76]: Learning something new! :)

[the remark this answers is not in the document]
Comment (iholder101) [reply tools/cache/time-defined-cache_test.go:76]: Yeah it surprised me as well :)
Good to know :smile:
